freqtrade/freqai/data_drawer.py

```diff
@@ -0,0 +1,371 @@
+            logger.info("Could not find existing historic_predictions, starting from scratch")
+        else:
+            logger.warning(
+                f"Follower could not find historic predictions at {self.full_path} "
+                "sending null values back to strategy"
+            )
+
+        return exists
+
+    def save_historic_predictions_to_disk(self):
+        """
+        Save data drawer full of all pair model metadata in present model folder.
+        """
+        with open(self.historic_predictions_path, "wb") as fp:
+            cloudpickle.dump(self.historic_predictions, fp, protocol=cloudpickle.DEFAULT_PROTOCOL)
+
+    def save_drawer_to_disk(self):
+        """
+        Save data drawer full of all pair model metadata in present model folder.
+        """
+        with open(self.pair_dictionary_path, "w") as fp:
+            json.dump(self.pair_dict, fp, default=self.np_encoder)
+
+    def save_follower_dict_to_disk(self):
+        """
+        Save follower dictionary to disk (used by strategy for persistent prediction targets)
+        """
+        with open(self.follower_dict_path, "w") as fp:
+            json.dump(self.follower_dict, fp, default=self.np_encoder)
+
+    def create_follower_dict(self):
+        """
+        Create or dictionary for each follower to maintain unique persistent prediction targets
+        """
+
+        whitelist_pairs = self.config.get("exchange", {}).get("pair_whitelist")
+
+        exists = self.follower_dict_path.is_file()
+
+        if exists:
+            logger.info("Found an existing follower dictionary")
+
+        for pair in whitelist_pairs:
+            self.follower_dict[pair] = {}
+
+        with open(self.follower_dict_path, "w") as fp:
+            json.dump(self.follower_dict, fp, default=self.np_encoder)
+
+    def np_encoder(self, object):
+        if isinstance(object, np.generic):
+            return object.item()
+
+    def get_pair_dict_info(self, pair: str) -> Tuple[str, int, bool, bool]:
+        """
+        Locate and load existing model metadata from persistent storage. If not located,
+        create a new one and append the current pair to it and prepare it for its first
+        training
+        :param pair: str: pair to lookup
+        :return:
+            model_filename: str = unique filename used for loading persistent objects from disk
+            trained_timestamp: int = the last time the coin was trained
+            coin_first: bool = If the coin is fresh without metadata
+            return_null_array: bool = Follower could not find pair metadata
+        """
+        pair_in_dict = self.pair_dict.get(pair)
+        data_path_set = self.pair_dict.get(pair, {}).get("data_path", None)
+        return_null_array = False
+
+        if pair_in_dict:
+            model_filename = self.pair_dict[pair]["model_filename"]
+            trained_timestamp = self.pair_dict[pair]["trained_timestamp"]
+            coin_first = self.pair_dict[pair]["first"]
```

Comment:
This segment could use some rework.
`pair_in_dict` actually contains the pair dict already (if the pair exists, obviously) - yet we're doing the dict lookup again several times below, which is ... subobtimal from a performance perspective.
i think there's two options - either you use `pair in self.pair_dict` to populate `pair_in_dict` (no need for dict lookup) and use a separte variable to cache the pair-dict - or reuse the variable within the if block.